### ISSUE TITLE
More description about how to store webhooks 

### ISSUE THREAD

**Issue by fotoflo** [Shopify/shopify-node-api#157]
## Overview/summary
from shopify-node-api/docs/usage/webhooks.md

> Note: The webhooks you register with Shopify are saved in the Shopify platform, but your handlers need to be reloaded whenever your server restarts. It is recommended to store your Webhooks in a persistent manner (for example, in a database) so that you can reload previously registered webhooks when your app restarts.

So what is it that needs to be stored? the webhook object that's returned here?

```
  const response = await Shopify.Webhooks.Registry.register({
    shop: config.shop,
    accessToken: config.accessToken,
    path: "/webhooks",
    topic: 'APP_UNINSTALLED',
    webhookHandler: async (topic, shop, body) =>{
      appReducer(shop, {action: 'APP_UNINSTALLED'})
    },
  });

```
In digging i found response.result.data.webhookSubscriptionCreate.webhookSubscription.id
it will return something like `gid://shopify/WebhookSubscription/12345678901234`

what do we do with this result?

Can you help update the documentation with a little more info about this?

...

## Motivation

> What inspired this enhancement?

learning more about webhooks

...

### Area

- [webhooks] Add any relevant `Area: <area>` labels to this issue
- [documentation]

---

## Checklist

- [ x ] I have described this enhancement in a way that is actionable (if possible)

**Comment by paulomarg** [comment]
Hey @fotoflo, great catch here. After that documentation was added, the `Shopify.Webhooks.Registry.register` method was updated to actually check if the webhook hasn't already been installed for a shop, so it can safely be run whenever an OAuth flow completes.

That also means you don't necessarily need to store the webhooks you'll be using - you just need to make sure to perform an OAuth after a server restart to load up the handlers.

You can still store and load your handlers just to make sure that the webhook registry is aware of the handlers you're expecting as soon as the server is up, if that is your preference. Since the handlers are the same for all shops, you'd actually only need to re-register them once per topic for any shop, so it's actually easier to just perform an OAuth. The purpose of reloading is just to have your app tell the webhook registry which handlers it should be aware of.

Hope this helps, but we'll update that documentation to recommend performing an OAuth rather than storing and loading the webhooks. Thanks for pointing it out.

**Comment by milancermak** [comment]
If we don't need to re-register webhook handlers between server restarts, since the local `webhookRegistry` won't be aware of any webhooks, won't it break the [process](https://github.com/Shopify/shopify-node-api/blob/main/docs/usage/webhooks.md#process-a-webhook) flow?

**Comment by paulomarg** [comment]
Hey @justoneabhi, as I mentioned, you don't need to re-register every shop, just one shop is enough. All we need is for a single call to `register` to be made, so that the library's registry knows that it should handle that topic.

That's why I suggested going through OAuth once with a shop you own, just so that it loads everything right away (without needing to wait for the first merchant to do a 'real' OAuth).

**Comment by paulomarg** [comment]
Yes and no, like I said the only thing that **needs** to happen is calling `register` once per topic, no matter how it happens. Doing an OAuth manually will do the trick if you register your handlers after OAuth, but it's not the only way to do it.

For example, if you want, you can persist (or hardcode) which topics you're using and what handlers to use for them and load them automatically on server startup, which is where the original (not ideally worded) suggestion in the docs came from. All you need to do that is a valid access token.

All that said, that is what the library currently allows. If there is a valid use case for doing something differently, we can look into it!

**Comment by milancermak** [comment]
First, I want to thank you @paulomarg for being so responsive and helpful in the discussions here on GitHub (not just this thread). I'm sure everyone greatly appreciates it 🙏

What's confusing to me about this whole webhooks business is that registering and handling of webhooks is tied to an app, however to actually register a handler, we need an OAuth token which is tied to a shop, and any shop at that. It would make much more sense if we could use the API key and secret to register webhook handlers, as those are tied to an app, not any particular shop.

FWIW, for now, I'm loading the offline token of my devshop from the database and using that to call `register` at every server (re)start, outside of any OAuth flow.

**Comment by paulomarg** [comment]
Hey @milancermak, that's a great point. I think we can extract the part of `register` that loads the handler function for a topic into a new method that does only that, so you can just load up your handlers without needing to actually hit Shopify.

**Comment by valorloff** [comment]
Great discussion!
My question is also about webhook :)
is there any communication between webhook ['app/uninstalled'] and session? 
Why Shopify.Utils.loadCurrentSession in '/webhook' ('app/uninstalled') route return undefined?

**Comment by shippinglob** [comment]
> First, I want to thank you @paulomarg for being so responsive and helpful in the discussions here on GitHub (not just this thread). I'm sure everyone greatly appreciates it 🙏 

Hi, I am developing my first Shopify app.  All the discussions I found on the issues of this repository were very useful to me. So thank you to @paulomarg for all the clarifications. And thanks also to all the app developers such as @milancermak who share the workarounds they come up with to bypass certain issues.

> FWIW, for now, I'm loading the offline token of my devshop from the database and using that to call `register` at every server (re)start, outside of any OAuth flow.

I liked the trick suggested here. I also read the code of the function `Shopify.Webhooks.Registry.register` 

My conclusion is:  if the webhook was previously registered on the Shopify side when calling `register` again for the same shop (`mustRegister=== false`), the only useful part of that function is the part that loads the hander. namely this line:

https://github.com/Shopify/shopify-node-api/blob/41e500bf865635ed92fbf7007afb52526564480b/src/webhooks/registry.ts#L223

By adding the following line, which does not require any token, at the beginning of my server file the handler seems to be loaded up properly when the server restarts and the events are correctly processed.

```node
Shopify.Webhooks.Registry.webhookRegistry.push({path: "/webhooks", topic: "ORDERS_PAID", webhookHandler: ordersPaidHandler});
```

If there is a use case where this would fail, please let me know 😊 

Otherwise, I would be happy to suggest a small PR that separates this line into a new "load" function that's exported from the module.

**Comment by jezsung** [comment]
Thanks for sharing @shippinglob @paulomarg. Registering webhooks when a server starts using `Shopify.Webhooks.Registry.webhookRegistry` looks great. I'm wondering if we should still need to register webhooks with `Shopify.Webhooks.Registry.register()` whenever a new shop completes OAuth flow. I don't know what `Shopify.Webhooks.Registry.register()` does under the hood with the `shop` and `accessToken`. Don't they handled and forwarded by `Shopify.Webhooks.Registry.process()`?

**Comment by bishpls** [comment]
If this is going to take time to be incorporated into the next API version, it may be worth updating the docs to show the current accepted "hack" method of handling server restarts? (https://github.com/Shopify/shopify-node-api/issues/157#issuecomment-841172584)

The Webhooks usage doc suggestion is pretty ambiguous / unhelpful here, in that it doesn't actually include the method of easily reloading via pushing to webhookRegistry: https://github.com/Shopify/shopify-node-api/blob/main/docs/usage/webhooks.md#register-a-webhook

"Note: The webhooks you register with Shopify are saved in the Shopify platform, but your handlers need to be reloaded whenever your server restarts. It is recommended to store your Webhooks in a persistent manner (for example, in a database) so that you can reload previously registered webhooks when your app restarts."

**Comment by jngbng** [comment]
I summarized this issue for me and someone who would be puzzled like me.

Before 2.0.0, `Shopify.Webhooks.Registry.register` api do two actions:
1. Register callback url to Shopify platform for shop & topic combination. (As far as I know, we cannot set a single static webhook url for all shops. Instead, we have to register webhook url for each shop. This setting is managed by Shopify.)
2. Dynamically register handler for the callback url to our server. `Shopify.Webhooks.Registry.process` uses this mapping to find the handler for a given topic. (this mapping is managed by our server.)

We have to keep {topic - handler} mapping across server reboot, but there is no public API for it and `Shopify.Webhooks.Registry.webhookRegistry.push(...)` is the only solution.

I found that as of 2.0.0, this issue is resolved by introducing `Shopify.Webhooks.Registry.addHandler`:
1. `Shopify.Webhooks.Registry.register` only registers callback url to Shopify platform. It does not require `webhookHandler` anymore.
2. programming has to register each handlers for topic using `Shopify.Webhooks.Registry.addHandler` when initializing the server.

**Comment by valorloff** [comment]
> 2\. A programmer has to register handlers for topic using `Shopify.Webhooks.Registry.addHandler` when initializing the server.

What is the difference between the new and the old approach? I don't see a fundamental difference

**Comment by jngbng** [comment]
> > 2. A programmer has to register handlers for topic using `Shopify.Webhooks.Registry.addHandler` when initializing the server.
> 
> What is the difference between the new and the old approach? I don't see a fundamental difference

@valorloff In the old approach, there was **no public method** to restore callback handler across server reboot.
`Shopify.Webhooks.Registry.webhookRegistry.push(...)` method is kind of hack exploiting internal implementation.
